### PR TITLE
Multi-input mv-lookup.

### PR DESCRIPTION
**Pull request by spherel**
These are the papers related to this lookup argument.
- multivariate lookup argument: https://eprint.iacr.org/2022/1530.
- cq: https://eprint.iacr.org/2022/1763

### CODE REVIEW

**Comment by Velaciela** [comment]
I did integration test with scroll-prover on the [mvlookup_integration](https://github.com/scroll-tech/halo2/commits/mvlookup_integration) branch, there are some compatibility issues with Scroll‘s proving system.

File: halo2_proofs/src/plonk/mv_lookup.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    // let degree = lhs_degree + 1
+    std::cmp::max(3, table_degree + 2)
+}
+
+pub fn degree_with_input(base_degree: usize, input_expression_degree: usize) -> usize {
+    base_degree + input_expression_degree
+}
+
+#[derive(Clone)]
+pub struct Argument<F: Field> {
```
Comment (Velaciela) [inline halo2_proofs/src/plonk/mv_lookup.rs:20]: [critical-1]
In [[lookup.rs]](https://github.com/geometryresearch/halo2/blob/82f4d6da44358fb07123f022120f510bd33ebb9c/halo2_proofs/src/plonk/lookup.rs#L9) we have `pub name: &'static str,`, mv_lookup missing this field, resulting compatibility issue. 
Temporarily fixed in [mv lookup Argument name](https://github.com/scroll-tech/halo2/commit/e7eeb464275194e8af97fec4a6cdd2f0ca90b75d), needs double-check to make sure all usage is consistent with the original lookup

File: halo2_proofs/src/plonk/mv_lookup.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    pub(crate) table_expressions: Vec<Expression<F>>,
+    pub(crate) inputs_expressions: Vec<Vec<Expression<F>>>,
```
Comment (Velaciela) [inline halo2_proofs/src/plonk/mv_lookup.rs:23]: [critical-2]
we use `let input = compress(lookup.input_expressions());`  in [snark-verifier](https://github.com/scroll-tech/snark-verifier/blob/4466059ce9a6dfaf26455e4ffb61d72af775cf52/snark-verifier/src/system/halo2.rs#L637) that expected `Vec<Expression<F>>` (not `Vec<Vec<Expression<F>>>`). During the integration test, i fake it by using  `let input = compress(lookup.table_expressions());`, so we can generate the proof, but can not pass the verification.

File: Cargo.lock [hunk 3]
```diff
@@ -4,9 +4,9 @@ version = 3
 
 [[package]]
 name = "addr2line"
-version = "0.17.0"
```
Comment (Velaciela) [inline Cargo.lock]: [minor-1]
probably no need to change `Cargo.lock`, I reverted it in [mvlookup_integration](https://github.com/scroll-tech/halo2/commits/mvlookup_integration) branch

File: halo2_proofs/src/plonk/evaluation.rs [hunk 4]
```diff
@@ -492,33 +508,127 @@ impl<C: CurveAffine> Evaluator<C> {
...
+                                            rot_scale,
+                                            isize,
+                                        )
+                                    })
+                                    .collect();
+
+                                inputs_values_for_extended_domain.extend_from_slice(&inputs_values);
+                            }
+
+                            inputs_values_for_extended_domain.batch_invert();
```
Comment (Velaciela) [inline halo2_proofs/src/plonk/evaluation.rs]: [perf-1]
[parallel batch invert](https://github.com/scroll-tech/halo2/commit/0b535ebe0f00d257eff3742b1550fb9a148c119d) saved 40*8=320s in phase4, better to have this optimization

File: halo2_proofs/src/plonk/circuit.rs [hunk 5]
```diff
@@ -1103,7 +1106,7 @@ impl<F: Field> Expression<F> {
     }
 }
 
-impl<F: std::fmt::Debug> std::fmt::Debug for Expression<F> {
+impl<F: ff::Field> std::fmt::Debug for Expression<F> {
```
Comment (Velaciela) [inline halo2_proofs/src/plonk/circuit.rs]: [critical-3]
their are some gadgets in zkevm-circuit e.g. [is_equal](https://github.com/scroll-tech/zkevm-circuits/blob/1093f5ecd2fee40fb7edb59df44bb30937a4e771/gadgets/src/is_equal.rs#L27-L34) has `std::fmt::Debug trait`, this change triggers compilation errors like:
```
28 | #[derive(Clone, Debug)]
| ----- in this derive macro expansion
...
33 | pub is_equal_expression: Expression<F>,
| ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `halo2_proofs::arithmetic::Field` is not implemented for `F`
|
= note: required for `Expression<F>` to implement `Debug`
```

**Review by Velaciela — COMMENTED** [review]
I did integration test with scroll-prover on the [mvlookup_integration](https://github.com/scroll-tech/halo2/commits/mvlookup_integration) branch, there are some compatibility issues with Scroll‘s proving system.

**Comment by Velaciela** [comment]
benchmarking on AWS g5.x12large machine (`48 vCore CPU, 192GB RAM`)
`scroll-prover (v0.5.10)` super circuit integration test (test_chunk_prove_verify)

note: the result have included parallel batch invert optimization


[mv-lookup] 
```
Start: phase2
End: phase2 ....................................................................66.325s
Start: phase3
End: phase3 ....................................................................275.776s
```
phase4 time total 2881.969s <> peak memory: 119GB, for each part:
```
··Start: coeff_to_extended_part
··End: coeff_to_extended_part ..................................................109.360s
··Start: custom_gates_time
··End: custom_gates_time .......................................................77.608s
··Start: permutations_time
··End: permutations_time .......................................................25.542s
··Start: lookup_time
····Start: inputs_inv_sum_time
····End: inputs_inv_sum_time ...................................................100.766s
····Start: lookups_iter_time (cosetfftpart and constraints calculation)
····End: lookups_iter_time .....................................................41.214s
··End: lookup_time .............................................................**141.981s**
```

---

[lookup]
```
Start: phase2
End: phase2 ....................................................................367.301s
Start: phase3
End: phase3 ....................................................................306.957s
```
phase4 time total: 2740.878s <> peak memory: 130GB, for each part:
```
··Start: coeff_to_extended_part
··End: coeff_to_extended_part ..................................................117.650s
··Start: custom_gates_time
··End: custom_gates_time .......................................................77.640s
··Start: permutations_time
··End: permutations_time .......................................................25.552s
··Start: lookup_time (cosetfftpart and constraints calculation)
··End: lookup_time .............................................................**120.937s**
```

File: halo2_gadgets/src/utilities/lookup_range_check.rs [hunk 6]
```diff
@@ -582,7 +582,6 @@ mod tests {
             assert_eq!(
                 prover.verify(),
                 Err(vec![VerifyFailure::Lookup {
-                    name: "lookup",
```
Comment (Velaciela) [inline halo2_gadgets/src/utilities/lookup_range_check.rs]: [critical-1] need this field back

File: halo2_proofs/src/plonk/mv_lookup.rs (hunk 2 above)
Comment (spherel) [reply halo2_proofs/src/plonk/mv_lookup.rs:23]: I think now the corresponding snark-verifier code for mvlookup hasn't been implemented.

**Comment by spherel** [comment]
> benchmarking on AWS g5.x12large machine (`48 vCore CPU, 192GB RAM`) `scroll-prover (v0.5.10)` super circuit integration test (test_chunk_prove_verify)
> 
> note: the result have included parallel batch invert optimization
> 
> [mv-lookup]
> 
> ```
> Start: phase2
> End: phase2 ....................................................................66.325s
> Start: phase3
> End: phase3 ....................................................................275.776s
> ```
> 
> phase4 time total 2881.969s <> peak memory: 119GB, for each part:
> 
> ```
> ··Start: coeff_to_extended_part
> ··End: coeff_to_extended_part ..................................................109.360s
> ··Start: custom_gates_time
> ··End: custom_gates_time .......................................................77.608s
> ··Start: permutations_time
> ··End: permutations_time .......................................................25.542s
> ··Start: lookup_time
> ····Start: inputs_inv_sum_time
> ····End: inputs_inv_sum_time ...................................................100.766s
> ····Start: lookups_iter_time (cosetfftpart and constraints calculation)
> ····End: lookups_iter_time .....................................................41.214s
> ··End: lookup_time .............................................................**141.981s**
> ```
> 
> [lookup]
> 
> ```
> Start: phase2
> End: phase2 ....................................................................367.301s
> Start: phase3
> End: phase3 ....................................................................306.957s
> ```
> 
> phase4 time total: 2740.878s <> peak memory: 130GB, for each part:
> 
> ```
> ··Start: coeff_to_extended_part
> ··End: coeff_to_extended_part ..................................................117.650s
> ··Start: custom_gates_time
> ··End: custom_gates_time .......................................................77.640s
> ··Start: permutations_time
> ··End: permutations_time .......................................................25.552s
> ··Start: lookup_time (cosetfftpart and constraints calculation)
> ··End: lookup_time .............................................................**120.937s**
> ```

**Phase2** computes `lookup.prepare`, including preparing the compressed input and table expressions and counting the occurrence of table entries in the input.
**Phase3** computes the running product of the permutation argument and the grand sum of the lookup argument.
**Phase4** prepares the quotient polynomial for the whole constraint system.

**Comment by spherel** [comment]
**MVLookup**
   MSM_COUNTER: {1048576: 1054}
   FFT_COUNTER: {1048576: 12431, 8388608: 1}

**Lookup**
   MSM_COUNTER: {1048576: 1652}
   FFT_COUNTER: {1048576: 17813, 8388608: 1}

**Comment by huwenqing0606** [comment]
a brief doc for this PR https://www.overleaf.com/read/vphfbrztgsmp, any comments/corrections would be greatly appreciated!

**Comment by kunxian-xia** [comment]
> benchmarking on AWS g5.x12large machine (`48 vCore CPU, 192GB RAM`) `scroll-prover (v0.5.10)` super circuit integration test (test_chunk_prove_verify)
> 
> note: the result have included parallel batch invert optimization
> 
> [mv-lookup]
> 
> ```
> Start: phase2
> End: phase2 ....................................................................66.325s
> Start: phase3
> End: phase3 ....................................................................275.776s
> ```
> 
> phase4 time total 2881.969s <> peak memory: 119GB, for each part:
> 
> ```
> ··Start: coeff_to_extended_part
> ··End: coeff_to_extended_part ..................................................109.360s
> ··Start: custom_gates_time
> ··End: custom_gates_time .......................................................77.608s
> ··Start: permutations_time
> ··End: permutations_time .......................................................25.542s
> ··Start: lookup_time
> ····Start: inputs_inv_sum_time
> ····End: inputs_inv_sum_time ...................................................100.766s
> ····Start: lookups_iter_time (cosetfftpart and constraints calculation)
> ····End: lookups_iter_time .....................................................41.214s
> ··End: lookup_time .............................................................**141.981s**
> ```
> 
> [lookup]
> 
> ```
> Start: phase2
> End: phase2 ....................................................................367.301s
> Start: phase3
> End: phase3 ....................................................................306.957s
> ```
> 
> phase4 time total: 2740.878s <> peak memory: 130GB, for each part:
> 
> ```
> ··Start: coeff_to_extended_part
> ··End: coeff_to_extended_part ..................................................117.650s
> ··Start: custom_gates_time
> ··End: custom_gates_time .......................................................77.640s
> ··Start: permutations_time
> ··End: permutations_time .......................................................25.552s
> ··Start: lookup_time (cosetfftpart and constraints calculation)
> ··End: lookup_time .............................................................**120.937s**
> ```

@Velaciela can you update the benchmark using the latest circuit (v0.9.3)?

**Comment by naure** [comment]
A summary of what happens for K lookups into the same table:
- Split K into batches K1, K2, etc, such that each batch respects the max constraint degree.

Then for each batch:
- Create a multiplicity column m.
- Create an accumulator.
- The constraint works on the difference between steps of the accumulator.

No helper columns are needed, but the batch size is limited by max degree.

Here is an optimisation:
- The first batch K1 works the same as above.
- But following batches do not need another instance of the entire argument.

Instead, for each following batch:
- Create a helper column containing inverses: `helper = 1 / (denom_0 * denom_1 * …)`
- A constraint verifies the divisions of this batch using the the helper column.
- The terms are added to the single accumulator. That does not increase its degree. The multiplicities are also added into the single m.

This way the multiplicity m and the accumulator are amortized across all lookups. With for example 4 lookups per helper columns, that's a cost of 0.25 columns per lookup + 1 shared multiplicity.

The accumulator counts as the helper of the first batch here. There is also the option to sacrifice this for simplicity; that is, use the same logic for all batches without the exception for the first batch. In that case, that would be 0.25 per lookup + 2.

(cc @spherel)

File: halo2_proofs/src/plonk/mv_lookup/prover.rs [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+        let m_values: Vec<F> = {
+            use std::sync::RwLock;
+            use std::sync::atomic::{AtomicU64, Ordering};
+            let m_values: Vec<AtomicU64> = (0..params.n()).map(|_| AtomicU64::new(0)).collect();
+
+            for compressed_input_expression in compressed_inputs_expressions.iter() {
+                compressed_input_expression.par_iter().take(params.n() as usize - blinding_factors - 1).for_each(|fi| {
+                    let index = table_index_value_mapping
+                        .get(fi)
+                        .unwrap();
```
Comment (kunxian-xia) [inline halo2_proofs/src/plonk/mv_lookup/prover.rs]: If some input `ci[j]` does not occur in the table, then it panics here which should instead return the `LookupError` back.

File: halo2_proofs/src/plonk/mv_lookup/prover.rs [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+    >(
+        self,
+        pk: &ProvingKey<C>,
+        params: &P,
+        beta: ChallengeBeta<C>,
+        mut rng: R,
+        transcript: &mut T,
+    ) -> Result<Committed<C>, Error> {
+        /*
+            φ_i(X) = f_i(X) + α
```
Comment (kunxian-xia) [inline halo2_proofs/src/plonk/mv_lookup/prover.rs:215]: The random challenge in comment does not match the input param (`α` vs. `beta`).

File: halo2_proofs/src/plonk/mv_lookup/prover.rs [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+                .iter()
+                .skip(params.n() as usize - blinding_factors)
+                .collect::<Vec<_>>();
+            assert_eq!(invalid_ms.len(), blinding_factors);
+            for mi in invalid_ms {
+                assert_eq!(*mi, C::Scalar::zero());
+            }
+
+            // check sums
+            let alpha = C::Scalar::random(&mut rng);
```
Comment (kunxian-xia) [inline halo2_proofs/src/plonk/mv_lookup/prover.rs:156]: It's better to rename this to `beta` to be consistent with the code below.

File: halo2_proofs/src/plonk/mv_lookup/prover.rs [hunk 10]
```diff
@@ -0,0 +1,413 @@
...
+
+        // Compute the evaluations of the lookup grand sum polynomial
+        // over our domain, starting with phi[0] = 0
+        let blinding_factors = pk.vk.cs.blinding_factors();
+        let phi = iter::once(C::Scalar::zero())
+            .chain(log_derivatives_diff)
+            .scan(C::Scalar::zero(), |state, cur| {
+                *state += &cur;
+                Some(*state)
+            })
```
Comment (kunxian-xia) [inline halo2_proofs/src/plonk/mv_lookup/prover.rs]: TODO: use parallel scan.

File: halo2_proofs/src/plonk/mv_lookup/prover.rs [hunk 11]
```diff
@@ -0,0 +1,413 @@
...
+            // Take all rows including the "last" row which should
+            // be a 0
+            .take(params.n() as usize - blinding_factors)
+            // Chain random blinding factors.
+            .chain((0..blinding_factors).map(|_| C::Scalar::random(&mut rng)))
+            .collect::<Vec<_>>();
+        assert_eq!(phi.len(), params.n() as usize);
```
Comment (kunxian-xia) [inline halo2_proofs/src/plonk/mv_lookup/prover.rs:313]: It's better to add one more check that `assert_eq!(phi[params.n() as usize - blinding_factors], F::zero());`

File: halo2_proofs/src/plonk/mv_lookup/verifier.rs [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+        advice_evals: &[C::Scalar],
+        fixed_evals: &[C::Scalar],
+        instance_evals: &[C::Scalar],
+        challenges: &[C::Scalar],
+    ) -> impl Iterator<Item = C::Scalar> + 'a {
+        let active_rows = C::Scalar::one() - (l_last + l_blind);
+
+        /*
+            φ_i(X) = f_i(X) + α
+            τ(X) = t(X) + α
```
Comment (kunxian-xia) [inline halo2_proofs/src/plonk/mv_lookup/verifier.rs]: Same as above (i.e. change `α` in comment to `beta` to match the code).

File: halo2_proofs/src/plonk/mv_lookup/verifier.rs [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+            .chain(
+                // phi[0] = 0
+                Some(l_0 * self.phi_eval),
+            )
+            .chain(
+                // phi[u] = 0
+                Some(l_last * self.phi_eval),
+            )
+            .chain(
+                // l_last(X) * (z(X)^2 - z(X)) = 0
```
Comment (kunxian-xia) [inline halo2_proofs/src/plonk/mv_lookup/verifier.rs]: The comment is wrong. It should be `(1 - l_last - l_blind) * (lhs - rhs) = 0`.

File: Cargo.lock (hunk 3 above)
Comment (kunxian-xia) [reply Cargo.lock]: Can you push the revert commit to this branch?

File: halo2_proofs/src/plonk/circuit.rs (hunk 5 above)
Comment (kunxian-xia) [reply halo2_proofs/src/plonk/circuit.rs]: Do not need to require `F: ff::Field`. Need to revert this change.

File: halo2_proofs/src/plonk/evaluation.rs [hunk 14]
```diff
@@ -534,41 +644,27 @@ impl<C: CurveAffine> Evaluator<C> {
...
+                                    inputs_prod * (table_value * inputs_inv_sum - m_coset[idx])
+                                };
+
+                                // phi[0] = 0
+                                *value = *value * y + l0[idx] * phi_coset[idx];
+
+                                // phi[u] = 0
+                                *value = *value * y + l_last[idx] * phi_coset[idx];
+
+                                // q(X) = LHS - RHS mod zH(X)
```
Comment (kunxian-xia) [inline halo2_proofs/src/plonk/evaluation.rs]: The comment here does not match the code.

File: halo2_proofs/src/plonk/evaluation.rs [hunk 15]
```diff
@@ -492,33 +508,127 @@ impl<C: CurveAffine> Evaluator<C> {
...
+                                .map(|input_lookup_evaluator| input_lookup_evaluator.instance())
+                                .collect();
+                            let mut table_eval_data = table_lookup_evaluator.instance();
+
                             for (i, value) in values.iter_mut().enumerate() {
                                 let idx = start + i;
 
-                                let table_value = lookup_evaluator.evaluate(
-                                    &mut eval_data,
+                                // f_i(X) + α for i in expressions
```
Comment (kunxian-xia) [inline halo2_proofs/src/plonk/evaluation.rs]: The comment is not accurate. It should be `beta` instead.

File: halo2_proofs/src/plonk/evaluation.rs (hunk 4 above)
Comment (kunxian-xia) [reply halo2_proofs/src/plonk/evaluation.rs]: @Velaciela can you push the change here?

File: halo2_proofs/src/plonk/mv_lookup/prover.rs (hunk 11 above)
Comment (spherel) [reply halo2_proofs/src/plonk/mv_lookup/prover.rs:313]: I think by `take(params.n() as usize - blinding_factors)`, `phi[params.n() as usize - blinding_factors]` is not included and it's a random element.

File: halo2_proofs/src/plonk/mv_lookup/prover.rs [hunk 16]
```diff
@@ -0,0 +1,440 @@
...
+
+        let blinding_factors = pk.vk.cs.blinding_factors();
+
+        // compute m(X)
+        let table_index_value_mapping: BTreeMap<C::Scalar, usize> = compressed_table_expression
+            .iter()
+            .take(params.n() as usize - blinding_factors - 1)
+            .enumerate()
+            .map(|(i, &x)| (x, i))
+            .collect();
```
Comment (kunxian-xia) [inline halo2_proofs/src/plonk/mv_lookup/prover.rs:116]: We need to run more benchmark on this op as constructing BTreeMap over a large table might take a lot of time.

**Comment by kunxian-xia** [comment]
Note that this pr  still has compilation errors. Even though, I will merge this pr into the `develop-rc` branch first as i don't have access to geometry's repo. Then i will make necessary changes there. 
cc @spherel

**Review by kunxian-xia — APPROVED** [review]

File: halo2_proofs/src/plonk/mv_lookup.rs (hunk 2 above)
Comment (kunxian-xia) [reply halo2_proofs/src/plonk/mv_lookup.rs:23]: Changes to `snark-verifier` is implemented in https://github.com/scroll-tech/snark-verifier/pull/24.

File: halo2_proofs/src/plonk/mv_lookup/prover.rs (hunk 16 above)
Comment (kunxian-xia) [reply halo2_proofs/src/plonk/mv_lookup/prover.rs:116]: This is fixed in #71 (commit https://github.com/scroll-tech/halo2/commit/d901014b088601b994a427a22ef280f217214b5a).

**Comment by AlekseiVambol** [comment]
> a brief doc for this PR https://www.overleaf.com/read/vphfbrztgsmp, any comments/corrections would be greatly appreciated!



> A summary of what happens for K lookups into the same table:
> 
>     * Split K into batches K1, K2, etc, such that each batch respects the max constraint degree.
> 
> 
> Then for each batch:
> 
>     * Create a multiplicity column m.
> 
>     * Create an accumulator.
> 
>     * The constraint works on the difference between steps of the accumulator.
> 
> 
> No helper columns are needed, but the batch size is limited by max degree.
> 
> Here is an optimisation:
> 
>     * The first batch K1 works the same as above.
> 
>     * But following batches do not need another instance of the entire argument.
> 
> 
> Instead, for each following batch:
> 
>     * Create a helper column containing inverses: `helper = 1 / (denom_0 * denom_1 * …)`
> 
>     * A constraint verifies the divisions of this batch using the the helper column.
> 
>     * The terms are added to the single accumulator. That does not increase its degree. The multiplicities are also added into the single m.
> 
> 
> This way the multiplicity m and the accumulator are amortized across all lookups. With for example 4 lookups per helper columns, that's a cost of 0.25 columns per lookup + 1 shared multiplicity.
> 
> The accumulator counts as the helper of the first batch here. There is also the option to sacrifice this for simplicity; that is, use the same logic for all batches without the exception for the first batch. In that case, that would be 0.25 per lookup + 2.
> 
> (cc @spherel)

Do I correctly understand that you propose to use one common constraint enforcing 
image 
where the sum is replaced by the sum of the polynomials A_i(x) found using interpolation, each of which corresponds to a separate batch and is constrained by something like this:
image
where K and f-polynomials a batch-specific?

**Comment by AlekseiVambol** [comment]
> a brief doc for this PR https://www.overleaf.com/read/vphfbrztgsmp, any comments/corrections would be greatly appreciated!

Good idea and nice paper draft! Two things can be done to make it even better:
1. It would be beneficient for people, who are not too familiar with transformation of the abstract polynomial protocols into the concrete ones by means of the polynomial commitments, to describe the motivation for transforming (4) into (5) by multiplying both sides by the same polynomial. If I am not mistaken, I would propose to describe this in the following way: 

"Some identity A(x) = B(x) over some domain, which in an abstract polynomial protocol would be checked by a trusted party, in the case of a concrete protocol based on the polynomial commitment is checked as follows:
- the prover computes Q(x) = (A(x) - B(x)) / Z(x), where Z(x) is a vanishing polynomial for that domain;
- the prover shares the commitments to the polynomials, which are the non-public parts of the expressions A(x) and B(x), and the commitment to Q(x);
- the verifier asks for opennings for all the commitments above at some random point p and checks whether the identity A(p) - B(p) = Q(p) *  Z(p) is correct. If the check yields the positive results, than the Schwarz-Zippel lemma tells the verifier that with the overwhelming probability A(x) = B(x) over that domain.
In order to apply this method, both A(x) and B(x) must be polynomials, and this is achieved by transforming (4) to (5). Actually, we even do not need to compute the coefficients of the right-hand side polynomial in (5), since we are interested only in Q(x), which can be found by means of FFT using only the values of the both sides and Z(x)."

2. The typo on the page 3 should be fixed: "1" has to be instead of "0" -
image

**Pull request merged**